Improved Age Stats gramplet

Nick-Hall opened the pull request:
* New Histogram widget
* Extra bucket for ages above maximum
* Automatic scaling to available width

romjerome commented:
:+1: Nice !

Maybe just a mistake on average and maximum ages?

![improved_age_stats](https://user-images.githubusercontent.com/10960439/67258173-abb86580-f48f-11e9-958e-23b718f41b92.png)

Also have a strange behavior during bulding (e.g., need to detach the gramplet),
but I have an old cairo version and this gramplet was never very fast on build anyway
under this configuration.

Great, thanks !

romjerome commented:
We can close #875

romjerome commented:
Do we need float for average on ages? 12 months on base 10!
```diff
-[_("Average"), glocale.format_string("%.1f", average)]
+[_("Average"), glocale.format_string("%.0f", average)]
```

romjerome commented:
I have strange haskeys on my data!

> [11, 12, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 65, 1814, 1818, 1820, 1821, 1822, 1825, 1826, 1828, 1829, 1830]

This might be related to dates using the french calendar ...
There is a config key for that.

romjerome commented:
```diff
--- agestats0.py	2019-10-22 06:35:16.933469608 +0200
+++ agestats.py	2019-10-22 06:33:24.353473929 +0200
@@ -120,12 +120,12 @@
             birth_date = None
             if birth_ref:
                 birth_event = self.dbstate.db.get_event_from_handle(birth_ref.ref)
-                birth_date = birth_event.get_date_object()
+                birth_date = birth_event.get_date_object().to_calendar(0)
             death_ref = p.get_death_ref()
             death_date = None
             if death_ref:
                 death_event = self.dbstate.db.get_event_from_handle(death_ref.ref)
-                death_date = death_event.get_date_object()
+                death_date = death_event.get_date_object().to_calendar(0)
             if (birth_date and birth_date.get_year() != 0 and
                 death_date and death_date.get_year() != 0):
                 age = death_date.get_year() - birth_date.get_year()
@@ -159,7 +159,7 @@
                         bref = f.get_birth_ref()
                         if bref:
                             bevent = self.dbstate.db.get_event_from_handle(bref.ref)
-                            bdate = bevent.get_date_object()
+                            bdate = bevent.get_date_object().to_calendar(0)
                             if (bdate and bdate.get_year() != 0 and
                                 birth_date and birth_date.get_year() != 0):
                                 diff = birth_date.get_year() - bdate.get_year()
@@ -175,7 +175,7 @@
                         bref = m.get_birth_ref()
                         if bref:
                             bevent = self.dbstate.db.get_event_from_handle(bref.ref)
-                            bdate = bevent.get_date_object()
+                            bdate = bevent.get_date_object().to_calendar(0)
                             if (bdate and bdate.get_year() != 0 and
                                 birth_date and birth_date.get_year() != 0):
                                 diff = birth_date.get_year() - bdate.get_year()
```

romjerome commented:
I do not know how to debug my display issue. I can reproduce it on dashboard and People views. Why building is better when the gramplet is detached ? I do not know ! 
Maybe a refresh issue on my screen?

About the maximum age and mixed calendars for dates calculation, we can reproduce the problem by using "example" database.

Thanks.

romjerome commented on gramps/plugins/gramplet/agestats.py:
```diff
@@ -30,273 +31,264 @@
+            event = self.dbstate.db.get_event_from_handle(ref.ref)
+            date = event.get_date_object()
+            if date.get_year() != 0:
+                return date
```
@Nick-Hall 

For dates using different calendars, maybe need
to force to one calendar before calculation?
```diff
return date.to_calendar('gregorian') # or .to_calendar('0') 
# or config.get('preferences.calendar-format-report')
```

prculley commented on gramps/plugins/gramplet/agestats.py:
```diff
@@ -30,273 +31,264 @@
+
+    def build_gui(self):
+        top = Gtk.ScrolledWindow()
+        top.set_policy(Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.AUTOMATIC)
```
You are putting a ScrolledWindow inside an already existing ScrolledWindow.  I suspect that this is causing the issue with redraw when first adding the Gramplet.  In any event when I removed that it seems to display properly.
<pre>
    def build_gui(self):
#         top = Gtk.ScrolledWindow()
#         top.set_policy(Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.AUTOMATIC)
        self.vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        self.vbox.set_margin_start(6)
        self.vbox.set_margin_end(6)
        self.vbox.set_margin_top(6)
        self.vbox.set_margin_bottom(6)
        self.vbox.set_spacing(12)
#         top.add(self.vbox)
        return self.vbox  #top
</pre>

prculley commented:
Closed my db and got:
2019-10-24 09:20:23.093: WARNING: dummydb.py: line 1427: database is closed
Continuing after gramplet error...
2019-10-24 09:20:23.093: WARNING: _gramplet.py: line 342: Gramplet gave an error: Age Stats
Traceback (most recent call last):
  File "C:\Users\prc\workspace\GrampsM\main\gramps\gen\plug\_gramplet.py", line 326, in _updater
    retval = next(self._generator)
  File "C:\Users\prc\workspace\GrampsM\main\gramps\plugins\gramplet\agestats.py", line 154, in main
    _("Age"), 5, self.max_age)
  File "C:\Users\prc\workspace\GrampsM\main\gramps\plugins\gramplet\agestats.py", line 253, in create_histogram
    if max(data) == 0:
ValueError: max() arg is an empty sequence

Probably should check for db closed in main...

romjerome commented:
:tada:  :1st_place_medal: 

It sounds good.
Thanks.

sam-m888 commented:
Visually very professional. 
- [ ] Issue with label resizing and how the charts looks on default start up for Gramps. Maybe ellipsis the end of the label(s) eg: "Lifespan Age Distri..."  when screen is too small.
![default size start - text issue ](https://user-images.githubusercontent.com/8924713/67609101-e34c4800-f7d6-11e9-87a2-3c58fd976bd8.png)

- [ ] Loss of ability to right click and copy the result to a email (Usage as mentioned a few times on the mailing list. (even I have done it)) Workaround is to possibly offer the existing version as a third party addon or work out a way so that when you copy it captures the text version  or outputs a chart image? 
![old age stats gramplet](https://user-images.githubusercontent.com/8924713/67609134-ffe88000-f7d6-11e9-8a43-b81b4e3c9f72.png)

- [ ] Update [manual](https://gramps-project.org/wiki/index.php/Gramps_5.1_Wiki_Manual_-_Gramplets#Age_Stats) (when rolled over for Gramps 5.2)

![dashboard](https://user-images.githubusercontent.com/8924713/67609369-1f33dd00-f7d8-11e9-8ff2-d5550b2e9f31.png)

Nick-Hall commented:
Updated.

* Improved date calculations using spans.  This should also fix the calendar issues.
* Fixed issue when database closed.
* Fixed nested `ScrolledWindow` bug.
* Ellipsize title when not enough space.

I'm still not sure how to copy to clipboard.  We could create image files.  Is it worth the effort?

wroldwiedbwe commented:
>> I'm still not sure how to copy to clipboard. We could create image files. Is it worth the effort?

Looks great 🥇 

My point of view, yesterday I copied the result to an email ( 4th time in the past few months!), while an image is ok, txt is preferred.  

How about including both gramplets as mentioned or is it possible to have an option or button or tab that lets you switch between the txt version and the image version ?

Nick-Hall commented:
@wroldwiedbwe Thanks.  We'll make the old version available for those that still want to use it.
